### PR TITLE
ARROW-9196: [C++][Compute] All casts accept scalar and sliced inputs

### PR DESCRIPTION
**Pull request by bkietz**
This is primarily a refactoring of the tests for "cast" to ensure that every case is also verified with scalar and sliced inputs through `CheckScalarUnary`. Caveat: ExtensionScalar is not implemented, so those checks aren't enabled. It should be straightforward for them to be enabled in a follow up, however.

NB: also resolves ARROW-9198

### CODE REVIEW

File: cpp/src/arrow/testing/gtest_util.h [hunk 1]
```diff
@@ -358,6 +354,14 @@ void ArrayFromVector(const std::vector<C_TYPE>& values, std::shared_ptr<Array>*
   ArrayFromVector<TYPE, C_TYPE>(type, values, out);
 }
 
+template <typename TYPE, typename C_TYPE = typename TYPE::c_type>
+std::shared_ptr<Array> ArrayFromVector(std::vector<bool> is_valid,
+                                       std::vector<C_TYPE> values) {
+  std::shared_ptr<Array> out;
+  ArrayFromVector<TYPE, C_TYPE>(is_valid, values);
+  return out;
```
Comment (pitrou) [inline cpp/src/arrow/testing/gtest_util.h:362]: Does this work? It looks like `out` isn't mutated in this function.

File: cpp/src/arrow/type_traits.h [hunk 2]
```diff
@@ -991,4 +991,20 @@ static inline bool is_nested(Type::type type_id) {
...
+static inline int offset_bit_width(Type::type type_id) {
+  switch (type_id) {
+    case Type::STRING:
+    case Type::BINARY:
+    case Type::LIST:
+      return 32;
+    case Type::LARGE_STRING:
+    case Type::LARGE_BINARY:
+    case Type::LARGE_LIST:
+      return 64;
```
Comment (pitrou) [inline cpp/src/arrow/type_traits.h:1003]: May want to add `MAP` and `DENSE_UNION` here?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 3]
```diff
@@ -475,1049 +207,821 @@ TEST_F(TestCast, CanCast) {
...
-  std::shared_ptr<Array> expected;
-  ArrayFromVector<Int16Type>(int16(), is_valid, e11, &expected);
-
-  auto buf = Buffer::Wrap(v11.data(), v11.size());
-  Int32Array tmp11(5, buf, expected->null_bitmap(), -1);
-
-  CheckPass(tmp11, *expected, int16(), options);
+TEST(Cast, OverflowInNullSlot) {
+  CheckCast(ArrayFromJSON(int32(), "[0, null, 2000, 1000, 0]"),
+            ArrayFromJSON(int16(), "[0, null, 2000, 1000, 0]"));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:263]: Did you forget to call `MaskArrayWithNullsAt` in this test?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 4]
```diff
@@ -1553,221 +1057,325 @@ TEST_F(TestCast, UnsupportedTargetType) {
...
-
-  options.allow_float_truncate = true;
-  CheckPass(*float64_list_array, *int32_list_array, int32_list_array->type(), options,
-            /*check_scalar=*/false);
-  CheckPass(*float64_list_array, *int64_list_array, int64_list_array->type(), options,
-            /*check_scalar=*/false);
+  CheckCast(list_int32, list_int64);
+  CheckCast(list_float32, list_int32);
+  CheckCast(list_int64, list_int32);
 }
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1350]: Can we add a simple test for `CastOptions` pass-through? e.g. a cast from `list<int32>` to `list<int16>` that would trigger overflow.

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 5]
```diff
@@ -1788,134 +1396,97 @@ TEST_F(TestCast, IdentityCasts) {
...
-    std::shared_ptr<Array> expected = *Take(*dict, *indices);
+  for (auto string_type : kBaseBinaryTypes) {
+    dictionaries.push_back(
+        ArrayFromJSON(string_type, R"(["foo", "bar", "baz", "foo", null])"));
+  }
 
-    this->CheckPass(*dict_arr, *expected, expected->type(), CastOptions::Safe(),
-                    /*check_scalar=*/false);
+  for (auto dict : dictionaries) {
+    for (auto index_type : all_dictionary_index_types()) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1460]: Nit: should we have `kDictionaryIndexTypes`?

**Review by pitrou — COMMENTED** [review]
Excellent! A bunch of comments below.

File: cpp/src/arrow/compute/kernels/scalar_cast_string.cc [hunk 6]
```diff
@@ -94,26 +94,45 @@ struct Utf8Validator {
 };
 
 template <typename I, typename O>
-struct CastBinaryToBinaryOffsets;
+static void CastBinaryToBinaryOffsets(KernelContext* ctx, const ArrayData& input,
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_string.cc:97]: yes, will remove

File: cpp/src/arrow/scalar.cc [hunk 7]
```diff
@@ -556,6 +556,12 @@ Status CastImpl(const ScalarType& from, StringScalar* to) {
   return Status::OK();
 }
 
+Status CastImpl(const Decimal128Scalar& from, StringScalar* to) {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/scalar.cc:559]: alright

File: cpp/src/arrow/testing/gtest_util.h [hunk 8]
```diff
@@ -44,16 +44,12 @@
...
-  do {                                                                                \
-    auto _res = (expr);                                                               \
-    ::arrow::Status _st = ::arrow::internal::GenericToStatus(_res);                   \
-    if (!_st.Is##ENUM()) {                                                            \
-      FAIL() << "Expected '" ARROW_STRINGIFY(expr) "' to fail with " ARROW_STRINGIFY( \
-                    ENUM) ", but got "                                                \
-             << _st.ToString();                                                       \
-    }                                                                                 \
-  } while (false)
+#define ASSERT_RAISES(ENUM, expr)                                                 \
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/testing/gtest_util.h:47]: will do

File: cpp/src/arrow/testing/gtest_util.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/testing/gtest_util.h:362]: Doesn't work; I started writing this overload and didn't end up using it. Will remove

File: cpp/src/arrow/type_traits.h (hunk 2 above)
Comment (bkietz) [reply cpp/src/arrow/type_traits.h:1003]: ah, good catch

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:263]: Will do

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 4 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1350]: will do

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 5 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:1460]: sure

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 3 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:263]: And now this test fails :-/

**Comment by bkietz** [comment]
@pitrou PTAL

**Review by pitrou — APPROVED** [review]
+1, thank you!

**Pull request closed without merging**